Leanplum-SDK/Classes/Managers/Networking/LPRequestSender.m

```diff
@@ -119,36 +120,45 @@ - (void)sendNow:(id<LPRequesting>)request sync:(BOOL)sync
             return;
         }
 
-        [self sendEventually:request];
+        [self sendEventually:request async:!sync];
         [self sendRequests:sync];
     }
     [self.countAggregator incrementCount:@"send_now_lp"];
 }
 
-- (void)sendEventually:(id<LPRequesting>)request
+- (void)sendEventually:(id<LPRequesting>)request async:(BOOL) async
 {
     if ([request isKindOfClass:[LeanplumRequest class]]) {
         LeanplumRequest *oldLeanplumRequest = request;
-        [oldLeanplumRequest sendEventually];
+        [oldLeanplumRequest sendEventually:async];
     } else {
         RETURN_IF_TEST_MODE;
         if (!request.sent) {
             request.sent = YES;
-            NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
-            NSString *uuid = [userDefaults objectForKey:LEANPLUM_DEFAULTS_UUID_KEY];
-            NSInteger count = [LPEventDataManager count];
-            if (!uuid || count % MAX_EVENTS_PER_API_CALL == 0) {
-                uuid = [self generateUUID];
-            }
 
-            @synchronized ([LPEventCallbackManager eventCallbackMap]) {
-                NSMutableDictionary *args = [self createArgsDictionaryForRequest:request];
-                args[LP_PARAM_UUID] = uuid;
-                [LPEventDataManager addEvent:args];
+            void (^operationBlock)(void) = ^void() {
+                NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
+                NSString *uuid = [userDefaults objectForKey:LEANPLUM_DEFAULTS_UUID_KEY];
+                NSInteger count = [LPEventDataManager count];
+                if (!uuid || count % MAX_EVENTS_PER_API_CALL == 0) {
+                    uuid = [self generateUUID];
+                }
+
+                @synchronized ([LPEventCallbackManager eventCallbackMap]) {
+                    NSMutableDictionary *args = [self createArgsDictionaryForRequest:request];
+                    args[LP_PARAM_UUID] = uuid;
+                    [LPEventDataManager addEvent:args];
+
+                    [LPEventCallbackManager addEventCallbackAt:count
+                                                     onSuccess:request.responseBlock
+                                                       onError:request.errorBlock];
+                }
+            };
 
-                [LPEventCallbackManager addEventCallbackAt:count
-                                                 onSuccess:request.responseBlock
-                                                   onError:request.errorBlock];
+            if (async) {
+                [[LPOperationQueue serialQueue] addOperationWithBlock:operationBlock];
```

Comment:
Still dont understand. This was anyways going to be part of batching call. So, why is this call itself put in Queue?

Reply:
This is not part of the batching, the `[LPEventDataManager addEvent:args];` will be run on main thread and will block it for the duration.